File: srcs/juloo.keyboard2/UserAdaptationManager.java
package juloo.keyboard2;

import android.content.Context;
import android.content.SharedPreferences;
import android.util.Log;
import java.util.HashMap;
import java.util.Map;

/**
 * Manages user adaptation by tracking word selection history and adjusting
 * word frequencies based on user preferences.
 */
public class UserAdaptationManager
{
  private static final String TAG = "UserAdaptationManager";
  private static final String PREFS_NAME = "user_adaptation";
  private static final String KEY_WORD_SELECTIONS = "word_selections_";
  private static final String KEY_TOTAL_SELECTIONS = "total_selections";
  private static final String KEY_LAST_RESET = "last_reset";
  
  // Configuration constants
  private static final int MIN_SELECTIONS_FOR_ADAPTATION = 5;
  private static final int MAX_TRACKED_WORDS = 1000;
  private static final float ADAPTATION_STRENGTH = 0.3f; // How much to boost frequently selected words
  private static final long RESET_PERIOD_MS = 30L * 24L * 60L * 60L * 1000L; // 30 days
  
  private final Context _context;
  private final SharedPreferences _prefs;
  private final Map<String, Integer> _selectionCounts;
  private int _totalSelections;
  private boolean _isEnabled;
  
  private static UserAdaptationManager _instance = null;
  
  public static synchronized UserAdaptationManager getInstance(Context context)
  {
    if (_instance == null)
    {
      _instance = new UserAdaptationManager(context.getApplicationContext());
    }
    return _instance;
  }
  
  private UserAdaptationManager(Context context)
  {
    _context = context;
    _prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
    _selectionCounts = new HashMap<>();
    _isEnabled = true;
    
    loadSelectionHistory();
    checkForPeriodicReset();
  }
  
  /**
   * Record that a word was selected by the user
   */
  public void recordSelection(String word)
  {
    if (!_isEnabled || word == null || word.trim().isEmpty())
      return;
    
    word = word.toLowerCase().trim();
    
    // Increment selection count
    int currentCount = _selectionCounts.getOrDefault(word, 0);
    _selectionCounts.put(word, currentCount + 1);
    _totalSelections++;
    
    // Limit the number of tracked words to prevent unbounded growth
    if (_selectionCounts.size() > MAX_TRACKED_WORDS)
    {
      pruneOldSelections();
    }
    
    // Save to persistent storage periodically (every 10 selections)
    if (_totalSelections % 10 == 0)
    {
      saveSelectionHistory();
    }
    
    Log.d(TAG, String.format("Recorded selection: '%s' (count: %d, total: %d)", 
                              word, currentCount + 1, _totalSelections));
  }
  
  /**
   * Get the adaptation multiplier for a word based on selection history
   * Returns 1.0 for no adaptation, >1.0 for frequently selected words
   */
  public float getAdaptationMultiplier(String word)
  {
    if (!_isEnabled || word == null || _totalSelections < MIN_SELECTIONS_FOR_ADAPTATION)
      return 1.0f;
    
    word = word.toLowerCase().trim();
    int selectionCount = _selectionCounts.getOrDefault(word, 0);
    
    if (selectionCount == 0)
      return 1.0f;
    
    // Calculate relative frequency (0 to 1)
    float relativeFrequency = (float) selectionCount / _totalSelections;
    
    // Apply adaptation strength to boost frequently selected words
    // Words selected often get up to 30% boost (with default ADAPTATION_STRENGTH)
    float multiplier = 1.0f + (relativeFrequency * ADAPTATION_STRENGTH * 10.0f);
    
    // Cap the maximum boost to prevent any single word from dominating
    multiplier = Math.min(multiplier, 2.0f);
    
    return multiplier;
  }
  
  /**
   * Get selection count for a specific word
   */
  public int getSelectionCount(String word)
  {
    if (word == null)
      return 0;
    return _selectionCounts.getOrDefault(word.toLowerCase().trim(), 0);
  }
  
  /**
   * Get total number of selections recorded
   */
  public int getTotalSelections()
  {
    return _totalSelections;
  }
  
  /**
   * Get number of unique words being tracked
   */
  public int getTrackedWordCount()
  {
    return _selectionCounts.size();
  }
  
  /**
   * Enable or disable user adaptation
   */
  public void setEnabled(boolean enabled)
  {
    _isEnabled = enabled;
    Log.d(TAG, "User adaptation " + (enabled ? "enabled" : "disabled"));
  }
  
  /**
   * Check if user adaptation is enabled
   */
  public boolean isEnabled()
  {
    return _isEnabled;
  }
  
  /**
   * Reset all adaptation data
   */
  public void resetAdaptation()
  {
    _selectionCounts.clear();
    _totalSelections = 0;
    
    // Clear from persistent storage
    SharedPreferences.Editor editor = _prefs.edit();
    editor.clear();
    editor.putLong(KEY_LAST_RESET, System.currentTimeMillis());
    editor.apply();
    
    Log.d(TAG, "User adaptation data reset");
  }
  
  /**
   * Get adaptation statistics for debugging
   */
  public String getAdaptationStats()
  {
    if (!_isEnabled)
      return "User adaptation disabled";
    
    StringBuilder stats = new StringBuilder();
    stats.append(String.format("User Adaptation Stats:\n"));
    stats.append(String.format("- Total selections: %d\n", _totalSelections));
    stats.append(String.format("- Unique words tracked: %d\n", _selectionCounts.size()));
    stats.append(String.format("- Adaptation active: %s\n", 
                               _totalSelections >= MIN_SELECTIONS_FOR_ADAPTATION ? "Yes" : "No"));
    
    if (_totalSelections >= MIN_SELECTIONS_FOR_ADAPTATION)
    {
      stats.append("\nTop 10 most selected words:\n");
      _selectionCounts.entrySet().stream()
        .sorted(Map.Entry.<String, Integer>comparingByValue().reversed())
        .limit(10)
        .forEach(entry -> {
          float multiplier = getAdaptationMultiplier(entry.getKey());
          stats.append(String.format("- %s: %d selections (%.2fx boost)\n", 
                                     entry.getKey(), entry.getValue(), multiplier));
        });
    }
    
    return stats.toString();
  }
  
  /**
   * Load selection history from persistent storage
   */
  private void loadSelectionHistory()
  {
    _totalSelections = _prefs.getInt(KEY_TOTAL_SELECTIONS, 0);
    
    // Load individual word counts
    Map<String, ?> allPrefs = _prefs.getAll();
    for (Map.Entry<String, ?> entry : allPrefs.entrySet())
    {
      String key = entry.getKey();
      if (key.startsWith(KEY_WORD_SELECTIONS))
      {
        String word = key.substring(KEY_WORD_SELECTIONS.length());
        if (entry.getValue() instanceof Integer)
        {
          _selectionCounts.put(word, (Integer) entry.getValue());
        }
      }
    }
    
    Log.d(TAG, String.format("Loaded adaptation data: %d total selections, %d unique words", 
                              _totalSelections, _selectionCounts.size()));
  }
  
  /**
   * Save selection history to persistent storage
   */
  private void saveSelectionHistory()
  {
    SharedPreferences.Editor editor = _prefs.edit();
    editor.putInt(KEY_TOTAL_SELECTIONS, _totalSelections);
    
    // Save individual word counts
    for (Map.Entry<String, Integer> entry : _selectionCounts.entrySet())
    {
      editor.putInt(KEY_WORD_SELECTIONS + entry.getKey(), entry.getValue());
    }
    
    editor.apply();
    
    Log.d(TAG, "Saved adaptation data to persistent storage");
  }
  
  /**
   * Remove least frequently selected words to prevent unbounded growth
   */
  private void pruneOldSelections()
  {
    // Find the minimum selection count threshold (remove bottom 20%)
    int targetSize = (int) (MAX_TRACKED_WORDS * 0.8);
    
    _selectionCounts.entrySet().stream()
      .sorted(Map.Entry.<String, Integer>comparingByValue())
      .limit(_selectionCounts.size() - targetSize)
      .map(Map.Entry::getKey)
      .forEach(_selectionCounts::remove);
    
    Log.d(TAG, String.format("Pruned selection data from %d to %d words", 
                              _selectionCounts.size() + (_selectionCounts.size() - targetSize), 
                              _selectionCounts.size()));
  }
  
  /**
   * Check if it's time for a periodic reset to prevent stale data
   */
  private void checkForPeriodicReset()
  {
    long lastReset = _prefs.getLong(KEY_LAST_RESET, System.currentTimeMillis());
    long timeSinceReset = System.currentTimeMillis() - lastReset;
    
    if (timeSinceReset > RESET_PERIOD_MS)
    {
      Log.d(TAG, "Performing periodic reset of adaptation data (30 days elapsed)");
      resetAdaptation();
    }
  }
  
  /**
   * Cleanup method to be called when the system is destroyed
   */
  public void cleanup()
  {
    saveSelectionHistory();
  }
}

File: srcs/juloo.keyboard2/UserAdaptationManager.kt
package juloo.keyboard2

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import kotlin.math.min

/**
 * Manages user adaptation by tracking word selection history and adjusting
 * word frequencies based on user preferences.
 */
class UserAdaptationManager private constructor(context: Context) {
    private val prefs: SharedPreferences
    private val selectionCounts: MutableMap<String, Int> = mutableMapOf()
    private var totalSelections: Int = 0
    private var isEnabled: Boolean = true

    init {
        prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        loadSelectionHistory()
        checkForPeriodicReset()
    }

    /**
     * Record that a word was selected by the user
     */
    fun recordSelection(word: String?) {
        if (!isEnabled || word.isNullOrBlank()) return

        val normalizedWord = word.lowercase().trim()

        // Increment selection count
        val currentCount = selectionCounts.getOrDefault(normalizedWord, 0)
        selectionCounts[normalizedWord] = currentCount + 1
        totalSelections++

        // Limit the number of tracked words to prevent unbounded growth
        if (selectionCounts.size > MAX_TRACKED_WORDS) {
            pruneOldSelections()
        }

        // Save to persistent storage periodically (every 10 selections)
        if (totalSelections % 10 == 0) {
            saveSelectionHistory()
        }

        Log.d(TAG, "Recorded selection: '$normalizedWord' (count: ${currentCount + 1}, total: $totalSelections)")
    }

    /**
     * Get the adaptation multiplier for a word based on selection history
     * Returns 1.0 for no adaptation, >1.0 for frequently selected words
     */
    fun getAdaptationMultiplier(word: String?): Float {
        if (!isEnabled || word == null || totalSelections < MIN_SELECTIONS_FOR_ADAPTATION) {
            return 1.0f
        }

        val normalizedWord = word.lowercase().trim()
        val selectionCount = selectionCounts.getOrDefault(normalizedWord, 0)

        if (selectionCount == 0) {
            return 1.0f
        }

        // Calculate relative frequency (0 to 1)
        val relativeFrequency = selectionCount.toFloat() / totalSelections

        // Apply adaptation strength to boost frequently selected words
        // Words selected often get up to 30% boost (with default ADAPTATION_STRENGTH)
        var multiplier = 1.0f + (relativeFrequency * ADAPTATION_STRENGTH * 10.0f)

        // Cap the maximum boost to prevent any single word from dominating
        multiplier = min(multiplier, 2.0f)

        return multiplier
    }

    /**
     * Get selection count for a specific word
     */
    fun getSelectionCount(word: String?): Int {
        if (word == null) return 0
        return selectionCounts.getOrDefault(word.lowercase().trim(), 0)
    }

    /**
     * Get total number of selections recorded
     */
    fun getTotalSelections(): Int = totalSelections

    /**
     * Get number of unique words being tracked
     */
    fun getTrackedWordCount(): Int = selectionCounts.size

    /**
     * Enable or disable user adaptation
     */
    fun setEnabled(enabled: Boolean) {
        isEnabled = enabled
        Log.d(TAG, "User adaptation ${if (enabled) "enabled" else "disabled"}")
    }

    /**
     * Check if user adaptation is enabled
     */
    fun isEnabled(): Boolean = isEnabled

    /**
     * Reset all adaptation data
     */
    fun resetAdaptation() {
        selectionCounts.clear()
        totalSelections = 0

        // Clear from persistent storage
        prefs.edit().apply {
            clear()
            putLong(KEY_LAST_RESET, System.currentTimeMillis())
            apply()
        }

        Log.d(TAG, "User adaptation data reset")
    }

    /**
     * Get adaptation statistics for debugging
     */
    fun getAdaptationStats(): String {
        if (!isEnabled) {
            return "User adaptation disabled"
        }

        val stats = StringBuilder()
        stats.append("User Adaptation Stats:\n")
        stats.append("- Total selections: $totalSelections\n")
        stats.append("- Unique words tracked: ${selectionCounts.size}\n")
        stats.append("- Adaptation active: ${if (totalSelections >= MIN_SELECTIONS_FOR_ADAPTATION) "Yes" else "No"}\n")

        if (totalSelections >= MIN_SELECTIONS_FOR_ADAPTATION) {
            stats.append("\nTop 10 most selected words:\n")
            selectionCounts.entries
                .sortedByDescending { it.value }
                .take(10)
                .forEach { (word, count) ->
                    val multiplier = getAdaptationMultiplier(word)
                    stats.append("- $word: $count selections (${"%.2f".format(multiplier)}x boost)\n")
                }
        }

        return stats.toString()
    }

    /**
     * Load selection history from persistent storage
     */
    private fun loadSelectionHistory() {
        totalSelections = prefs.getInt(KEY_TOTAL_SELECTIONS, 0)

        // Load individual word counts
        val allPrefs = prefs.all
        for ((key, value) in allPrefs) {
            if (key.startsWith(KEY_WORD_SELECTIONS)) {
                val word = key.substring(KEY_WORD_SELECTIONS.length)
                if (value is Int) {
                    selectionCounts[word] = value
                }
            }
        }

        Log.d(TAG, "Loaded adaptation data: $totalSelections total selections, ${selectionCounts.size} unique words")
    }

    /**
     * Save selection history to persistent storage
     */
    private fun saveSelectionHistory() {
        prefs.edit().apply {
            putInt(KEY_TOTAL_SELECTIONS, totalSelections)

            // Save individual word counts
            for ((word, count) in selectionCounts) {
                putInt(KEY_WORD_SELECTIONS + word, count)
            }

            apply()
        }

        Log.d(TAG, "Saved adaptation data to persistent storage")
    }

    /**
     * Remove least frequently selected words to prevent unbounded growth
     */
    private fun pruneOldSelections() {
        // Find the minimum selection count threshold (remove bottom 20%)
        val targetSize = (MAX_TRACKED_WORDS * 0.8).toInt()
        val originalSize = selectionCounts.size

        val wordsToRemove = selectionCounts.entries
            .sortedBy { it.value }
            .take(selectionCounts.size - targetSize)
            .map { it.key }

        wordsToRemove.forEach { selectionCounts.remove(it) }

        Log.d(TAG, "Pruned selection data from $originalSize to ${selectionCounts.size} words")
    }

    /**
     * Check if it's time for a periodic reset to prevent stale data
     */
    private fun checkForPeriodicReset() {
        val lastReset = prefs.getLong(KEY_LAST_RESET, System.currentTimeMillis())
        val timeSinceReset = System.currentTimeMillis() - lastReset

        if (timeSinceReset > RESET_PERIOD_MS) {
            Log.d(TAG, "Performing periodic reset of adaptation data (30 days elapsed)")
            resetAdaptation()
        }
    }

    /**
     * Cleanup method to be called when the system is destroyed
     */
    fun cleanup() {
        saveSelectionHistory()
    }

    companion object {
        private const val TAG = "UserAdaptationManager"
        private const val PREFS_NAME = "user_adaptation"
        private const val KEY_WORD_SELECTIONS = "word_selections_"
        private const val KEY_TOTAL_SELECTIONS = "total_selections"
        private const val KEY_LAST_RESET = "last_reset"

        // Configuration constants
        private const val MIN_SELECTIONS_FOR_ADAPTATION = 5
        private const val MAX_TRACKED_WORDS = 1000
        private const val ADAPTATION_STRENGTH = 0.3f // How much to boost frequently selected words
        private const val RESET_PERIOD_MS = 30L * 24L * 60L * 60L * 1000L // 30 days

        @Volatile
        private var instance: UserAdaptationManager? = null

        @JvmStatic
        fun getInstance(context: Context): UserAdaptationManager {
            return instance ?: synchronized(this) {
                instance ?: UserAdaptationManager(context.applicationContext).also { instance = it }
            }
        }
    }
}
